Mathlib/Topology/SubsetProperties.lean

```diff
@@ -1307,6 +1309,120 @@ theorem IsClosed.exists_minimal_nonempty_closed_subset [CompactSpace α] {S : Se
+    (hcomp : ∀ (s : Set α), s ∈ S → IsCompact s) : IsSigmaCompact (⋃₀ S) := by
+  by_cases S = ∅
+  · simp only [h, sUnion_empty, isSigmaCompact_empty]
+  · -- If S is non-empty, choose a surjection f : ℕ → S, this yields a map ℕ → Set X.
+    obtain ⟨f, hf⟩ := (Set.countable_iff_exists_surjective (nmem_singleton_empty.mp h)).mp hc
+    refine ⟨fun n ↦ f n, fun n ↦ hcomp (f n) (Subtype.mem (f n)), ?_⟩
+    -- I presume this part can be golfed.
+    apply Subset.antisymm
+    · -- Suppose x ∈ ⋃ n, f n. Then x ∈ f i for some i,
+      intro _ hx
+      rw [mem_iUnion] at hx
+      rcases hx with ⟨i, hi⟩
+      -- ... but f i is a set in S, so x ∈ ⋃ S as well.
+      exact ⟨f i, Subtype.mem (f i), hi⟩
+    · -- Suppose x ∈ ⋃ s, then x ∈ s for some s ∈ S.
+      intro x hx
+      rw [mem_sUnion] at hx
+      rcases hx with ⟨s, h, hxs⟩
+      -- Choose n with f n = s (using surjectivity of f).
+      have : ∃ n, f n = s := by
+        obtain ⟨y, hy⟩ := hf ⟨s, h⟩
+        use y
+        simp_all only [Subtype.forall]
+      rcases this with ⟨n, hn⟩
+      exact ⟨f n, mem_range_self n, (by rw [hn]; exact hxs)⟩
+
+/-- Countable unions of σ-compact sets are σ-compact. -/
+lemma isSigmaCompact_of_countable_sigma_compact (S : Set (Set α)) (hc : Set.Countable S)
+    (hcomp : ∀ s : S, IsSigmaCompact s (α := α)) : IsSigmaCompact (⋃₀ S) := by
+  -- Choose a decomposition s = ⋃ s_i for each s ∈ S.
+  choose K hcomp hcov using fun s ↦ hcomp s
+  -- Then, we have a countable union of countable unions of compact sets, i.e. countably many.
+  have : ⋃₀ S = ⋃₀ range (K.uncurry) := by
+    calc ⋃₀ S
+      _ = ⋃ s : S, s := by rw [← sUnion_eq_iUnion]
+      _ = ⋃ s : S, ⋃ n, (K s n) := by simp_rw [hcov]
+      _ = ⋃ s : S, ⋃ n, (K.uncurry ⟨s, n⟩) := by rw [Function.uncurry_def]
+      _ = ⋃ (s : S) (n : ℕ), (K.uncurry ⟨s, n⟩) := by rw [iUnion_coe_set]
+      _ = ⋃₀ range (K.uncurry) := by rw [iUnion_prod', sUnion_range]
+  rw [this]
+  rw [← countable_coe_iff] at hc
+  refine isSigmaCompact_of_countable_compact _ (countable_range (K.uncurry)) fun s hs ↦ ?_
+  obtain ⟨⟨ys, yn⟩, hy⟩ := mem_range.mp hs
+  rw [← hy]
+  exact hcomp ys yn
+
+/-- A closed subset of a σ-compact set is σ-compact. -/
+lemma IsSigmaCompact.of_isClosed_subset {s t : Set α} (ht : IsSigmaCompact t)
+    (hs : IsClosed s) (h : s ⊆ t) : IsSigmaCompact s := by
+  rcases ht with ⟨K, hcompact, hcov⟩
+  refine ⟨(fun n ↦ s ∩ (K n)), fun n ↦ (hcompact n).inter_left hs, ?_⟩
+  rw [← inter_iUnion, hcov]
+  exact Set.inter_eq_left_iff_subset.mpr h
+
+/-- If `s` is σ-compact and `f` continuous on a set `w` containing `s`, `f(s)` is σ-compact.-/
+lemma IsSigmaCompact.image_of_continuousOn {f : α → β} {s : Set α} (hs : IsSigmaCompact s)
+    (hf : ContinuousOn f s) : IsSigmaCompact (f '' s) := by
+  rcases hs with ⟨K, hcompact, hcov⟩
+  refine ⟨fun n ↦ f '' K n, ?_, (by rw [← hcov]; exact image_iUnion.symm)⟩
+  refine fun n ↦ (hcompact n).image_of_continuousOn (hf.mono ?_)
+  calc K n
```

Comment:
Would changing this calc block to
```
  rw [← hcov]
  exact subset_iUnion K n
```
be desirable? (My sense of aesthetics is unsure.)

Reply:
I think it's inconsistent to use a calc block for this one but `by rw [← hcov]; exact image_iUnion.symm` for the other one.

Reply:
Fair point, agreed.